### DISCUSSION TITLE
Concurrency design

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#89] [Ideas]
Looking for ideas on how to approach concurrency. My experience is limited to some MPI work back in my C/C++ days.

Things we can do in parallel include:
- traversing and updating shape points in procedural modeling
- procedural generation of shapes
- running animators
- building render lists for display processing

What are some ways to do this? Clojure-like facilities? Bordeaux threads? Thread pooling?

Any clean and simple libraries we should consider?

**Comment by jolby** [comment]
lparallel is probably worth considering:
https://github.com/lmj/lparallel

**Reply by mikelevins** [reply]
I've used it somewhat extensively in discrete event simulations in my day job. It works well if you can structure things as several long-running threads that communicate by passing messages. My code generally treats the messages as events. It queues them using concurrent queues and dispatches them according to computed discriminators to event handlers.

It uses a kernel that manages threads and communication, and which you have to setup and manage, so it's not just some added library calls. You also have to manage kernel state appropriately.

You can distribute it across hosts using lfarm, but we haven't used that, yet.

**Reply by ghost** [reply]
Is lparallel similar to the model in Concurrent ML?  Some Scheme implementations have similar functionality, Scheme48, Guile, and Racket for example.

https://wingolog.org/archives/2017/06/29/a-new-concurrent-ml

**Comment by kaveh808** [comment]
I also came across this:

https://github.com/bpecsek/spectralnorm4/blob/main/src/spectralnorm4.lisp

**Comment by ghost** [comment]
This *is* a good topic to explore up front.

It is complicated by a number of things.  In a stand-alone desktop-style application, you have a multicore CPU and multicore GPU most likely.  Vendors are starting to add machine learning coprocessors, which can also be useful for graphics work.  There is some lock-in with what technology you choose to coordinate between the GPU and CPU.  Should you let that lock you in to what is to come?

Scaling to a cloud or server farm, which is conceivable if you are doing a full-scale production with kons-9 or doing mobile/web SaaS, adds additional complexity--single threaded unikernels, containers, VMs, etc.  Is each node in a unikernel solution going to run a 30MB Common Lisp instance?

Even if you scope the initial implementation of kons-9 to desktop only, it does make sense to think about the end-goal of such a system.  My own general opinion is to keep things orthogonal and composable.

**Reply by kaveh808** [reply]
Though the idea of a desktop environment that can be extended by cloud and server farms is attractive, let's just consider the desktop-style setup for now.

Traditionally in production people go out to server farms to run large batch processes. Simulations, rendering, and the like.
